### PR TITLE
Add integration tests for REST and messaging exporters

### PR DESCRIPTION
**Pull request by srnagar**
This PR adds integration tests for testing the exporter with:
- REST client (Using App Configuration)
- Messaging client  (Using Event Hubs producer)
- Processor client (Using Event Processor Client for Event Hubs)

Testing the exporter in isolation already exists and is not part of this PR.

The test plan is detailed in this doc - https://gist.github.com/pakrym/8902e50db6cb9ce431083ad461dae2cc

image

### CODE REVIEW

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/pom.xml [hunk 1]
```diff
@@ -34,7 +34,7 @@
     <dependency>
       <groupId>com.azure</groupId>
       <artifactId>azure-core</artifactId>
-      <version>1.12.0</version> <!-- {x-version-update;com.azure:azure-core;dependency} -->
+      <version>1.13.0-beta.1</version> <!-- {x-version-update;com.azure:azure-core;dependency} -->
```
Comment (alzimmermsft) [inline sdk/monitor/azure-opentelemetry-exporter-azuremonitor/pom.xml:37]: What was added that made this change to an unreleased dependency? If this is remaining an unreleased dependency you'll need to update the version tag

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AppConfigurationExporterIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+public class AppConfigurationExporterIntegrationTest extends AzureMonitorExporterTestBase {
+
+    @Test
+    public void setConfigurationTest() throws InterruptedException {
+        CountDownLatch appConfigCountDown = new CountDownLatch(1);
+        CountDownLatch exporterCountDown = new CountDownLatch(2);
+
+        ConfigurationClient client = getConfigurationClient(appConfigCountDown);
+        Tracer tracer = configureAzureMonitorExporter((context, next) -> {
+            Mono<String> asyncString = FluxUtil.collectBytesInByteBufferStream(context.getHttpRequest().getBody())
```
Comment (alzimmermsft) [inline sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AppConfigurationExporterIntegrationTest.java:30]: Watch out on FluxUtil consuming the request body to completion. `collectByesInByteBufferStream` will mutate the underlying `ByteBuffer`s' `remaining()` value which could cause `hasRemaining()` to be false when the networking layer attempts to send the request

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/main/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporter.java [hunk 3]
```diff
@@ -406,6 +407,7 @@ private void exportRequest(String stdComponent, SpanData span, List<TelemetryIte
         }
 
         Double samplingPercentage = removeAiSamplingPercentage(attributes);
+        samplingPercentage = samplingPercentage == null ? 100.0 : samplingPercentage;
```
Comment (samvaity) [inline sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/main/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporter.java:410]: Is/Should this be defaulted by the exporter when provided null?

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporterTestBase.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    public void teardownTest(TestInfo testInfo) {
+    }
+
+    Tracer configureAzureMonitorExporter(HttpPipelinePolicy validator) {
+
+        HttpClient httpClient;
+        if (getTestMode() == TestMode.RECORD || getTestMode() == TestMode.LIVE) {
+            httpClient = HttpClient.createDefault();
+        } else {
+            httpClient = interceptorManager.getPlaybackClient();
```
Comment (samvaity) [inline sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporterTestBase.java:35]: Are we doing recordings for these tests?

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporterTestBase.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        }
+
+        HttpPipeline httpPipeline = new HttpPipelineBuilder()
+            .httpClient(httpClient)
+            .policies(interceptorManager.getRecordPolicy(), validator).build();
+
+        AzureMonitorExporter exporter = new AzureMonitorExporterBuilder()
+            .connectionString("{connection-string}")
```
Comment (samvaity) [inline sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporterTestBase.java:43]: If recording, should these be looking for env vars?

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/EventHubsExporterIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        producer.createBatch(new CreateBatchOptions().setPartitionId("0"))
+            .flatMap(batch -> {
+                batch.tryAdd(new EventData("test event "));
+                return producer.send(batch);
+            }).block();
+
+        assertTrue(eventCountDown.await(10, TimeUnit.SECONDS));
+        assertTrue(exporterCountDown.await(10, TimeUnit.SECONDS));
+    }
+}
```
Comment (samvaity) [inline sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/EventHubsExporterIntegrationTest.java:121]: Do we need a comprehensive test for this https://gist.github.com/pakrym/8902e50db6cb9ce431083ad461dae2cc#all-working-together

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/pom.xml (hunk 1 above)
Comment (srnagar) [reply sdk/monitor/azure-opentelemetry-exporter-azuremonitor/pom.xml:37]: Had to take a dependency on unreleased `azure-core-tracing-opentelemetry` which in turn depends on unreleased core (current).

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporterTestBase.java (hunk 4 above)
Comment (srnagar) [reply sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporterTestBase.java:35]: Currently, this is set up as an integration test only. So, recordings are not included.

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporterTestBase.java (hunk 5 above)
Comment (srnagar) [reply sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporterTestBase.java:43]: Live test setup will be added as a follow-up.

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/EventHubsExporterIntegrationTest.java (hunk 6 above)
Comment (srnagar) [reply sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/EventHubsExporterIntegrationTest.java:140]: Added storage checkpoint store which will cover the comprehensive test scenario.

**Review by joshfree — APPROVED** [review]

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/main/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporter.java (hunk 3 above)
Comment (srnagar) [reply sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/main/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporter.java:410]: Sampling will be completely removed as it won't be supported in this exporter.

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AppConfigurationExporterIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    private ConfigurationClient getConfigurationClient(CountDownLatch appConfigCountDown) {
+        ConfigurationClient client = new ConfigurationClientBuilder()
+            .connectionString(System.getenv("APP_CONFIG_CONNECTION_STRING"))
+            .addPolicy((context, next) -> {
+                Optional<Object> data = context.getData(com.azure.core.util.tracing.Tracer.AZ_TRACING_NAMESPACE_KEY);
+                if (data.isPresent() && data.get().equals("Microsoft.AppConfiguration")) {
+                    appConfigCountDown.countDown();
+                }
```
Comment (trask) [inline sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AppConfigurationExporterIntegrationTest.java:66]: just curious, what is this data attached to the appconfiguration requests used for?

**Review by trask — APPROVED** [review]

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/main/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporter.java [hunk 8]
```diff
@@ -349,6 +355,7 @@ private void exportRequest(String stdComponent, SpanData span, List<TelemetryIte
         addLinks(requestData.getProperties(), span.getLinks());
         Long httpStatusCode = attributes.get(SemanticAttributes.HTTP_STATUS_CODE);
 
+        requestData.setResponseCode("200");
```
[the remark this answers is not in the document]
Comment (srnagar) [reply sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/main/java/com/azure/opentelemetry/exporter/azuremonitor/AzureMonitorExporter.java:356]: Yeah, this was required as the response code was required field to send to service.

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AppConfigurationExporterIntegrationTest.java (hunk 7 above)
Comment (srnagar) [reply sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/AppConfigurationExporterIntegrationTest.java:66]: The data is just a property bag (Context) associated with the HTTP request which can be used for processing the request in the HTTP pipeline. This is used for setting some flags to add some hints to request processing.

File: sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/InMemoryCheckpointStore.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (srnagar) [reply sdk/monitor/azure-opentelemetry-exporter-azuremonitor/src/test/java/com/azure/opentelemetry/exporter/azuremonitor/InMemoryCheckpointStore.java:1]: Good idea! I will add another integration test for testing this flag in the follow-up PR.

**Pull request merged**